images/hook/entrypoint.sh

```diff
@@ -1,44 +1,112 @@
-    #               yet installed and thus just creates all OpenEBS resources
-    #               from scratch.
-    #               In the future, we'll need to develop a proper upgrade procedure
-    #               for OpenEBS components and its pools/volumes like described in
-    #               https://github.com/openebs/openebs/tree/master/k8s/upgrades.
+  CONTROL_PLANE_COMPONENTS=control_plane_components.txt
+  kubectl get pods -n openebs -l openebs.io/version="$1" > $CONTROL_PLANE_COMPONENTS
 
-    echo "--> Creating OpenEBS resources"
-    rig upsert -f /var/lib/gravity/resources/openebs-operator.yaml --debug
+  echo "Found control plane components:"
+  cat $CONTROL_PLANE_COMPONENTS
 
-    echo "--> Checking status"
-    rig status ${RIG_CHANGESET} --retry-attempts=120 --retry-period=1s --debug
+  grep 'provisioner.*Running' $CONTROL_PLANE_COMPONENTS >/dev/null &&
+    grep 'admission-server.*Running' $CONTROL_PLANE_COMPONENTS >/dev/null &&
+    grep 'maya-apiserver.*Running' $CONTROL_PLANE_COMPONENTS >/dev/null &&
+    grep 'ndm.*Running' $CONTROL_PLANE_COMPONENTS >/dev/null &&
+    grep 'snapshot-operator.*Running' $CONTROL_PLANE_COMPONENTS >/dev/null
 
-    echo "--> Freezing"
-    rig freeze
+  return $?
+}
 
-elif [ $1 = "rollback" ]; then
+if [ "$1" = "update" ]; then
+  TO_VERSION=2.2.0
 
-    echo "--> Reverting changeset $RIG_CHANGESET"
-    rig revert
-    rig cs delete --force -c cs/${RIG_CHANGESET}
+  echo "--> Checking: $RIG_CHANGESET"
+  if rig status "${RIG_CHANGESET}" --retry-attempts=1 --retry-period=1s; then
+    exit 0
+  fi
 
-else
+  echo "--> Starting upgrade, changeset: $RIG_CHANGESET"
+
+  echo "Starting the control plane upgrade process as described at:"
+  echo "https://github.com/openebs/openebs/blob/master/k8s/upgrades/README.md"
+
+  echo "Checking the existing control plane version..."
+  FROM_VERSION=$(get_control_plane_version)
+  if [[ "$FROM_VERSION" =~ ^[0-9]+.[0-9]+.[0-9]+$ ]]; then
+    if [ "$FROM_VERSION" == "$TO_VERSION" ]; then
+      echo "The control plane is already upgraded TO_VERSION=$TO_VERSION."
+    else
+      if [ "$FROM_VERSION" == "1.4.0" ] || [ "$FROM_VERSION" == "1.5.0" ] || [ "$FROM_VERSION" == "1.7.0" ]; then
+        echo "Found control plane components of expected FROM_VERSION=$FROM_VERSION."
+      else
+        echo "Exiting because unable to find expected control plane components FROM_VERSION. Got: '$FROM_VERSION'."
+        exit 1
+      fi
+    fi
+  else
+    echo "Exiting because unable to retrieve existing control plane version. Got: '$FROM_VERSION'."
+    exit 2
+  fi
+
+  if [ "$FROM_VERSION" != "$TO_VERSION" ]; then
+    echo "Performing control plane upgrade TO_VERSION=$TO_VERSION..."
+    if ! rig upsert -f /var/lib/gravity/resources/openebs-operator.yaml --debug;
+    then
+      echo "Failed rig upsert openebs-operator. Exiting."
+      exit $?
+    fi
 
-    echo "--> Missing argument, should be either 'update' or 'rollback'"
+    for i in {1..7}; do
```

Comment:
This check looks like it might be redundant with what rigging is already doing. Rigging I believe checks dependent pods are scheduled and running before completing. So when creating a Daemonset for example, it checks that the Daemonset is Running on all nodes before completing.